Working Natively

downpat opened the pull request:
Hey Chris and Doug,

I was able to get it working natively. Ubuntu's native node uses "nodejs" as the binary, so I had to change the package.json script to reflect. I also had to update the webpack defaults to work with my headless Ubuntu Server VM, so I could access it from the browser. After I did that, it seemed like there was a capitalization error in the React component, so I fixed that too. I think I'm ready to handle the kata now. You can safely ignore this pull request.

-Updated package.json to work with native node on ubuntu
-Updated .gitignore to ignore .swp files created by vim
-Updated todo_list.js to work with todo.js

downpat committed:
-Updated package.json to work with native node on ubuntu
-Updated .gitignore to ignore .swp files created by vim
-Updated todo_list.js to work with todo.js

dmorgan3405 commented:
Awesome! Sorry was unable to spend enough time today to get the hot swapping working. 

So, as far as what we would like to see. Simply add the functionality to allow a user to add a todo to the list. 

No need to worry about persistence. 

If you have any questions please reach out!

dmorgan3405 commented:
@stollcri just looping you in.

downpat commented:
@dmorgan3405  and @stollcri,

I spent some time working through this, but adding a todo list item to a React/Redux application is beyond my abilities at the moment. Here's a summary of the work I did today. If you would like me to continue tomorrow, I can.

- Got the app working in my native environment
- Fixed a typo that was causing the app to not render correctly in the browser
- Tried to get the unit tests working, first natively, then using the new Dockerfile that Doug added
- Added a docker compose file to work directly on the image created by the dockerfile
- Found that the unit tests weren't working because they require DOM, but moved on because it wasn't part of the task anyhow
- Read through this tutorial: https://facebook.github.io/react/tutorial/tutorial.html and got to know how React components work
- Used grep and other linux utilities to get to know Actions and Reducers, then went to the Redux documentation and read about them (http://redux.js.org/docs/basics/Actions.html  &  http://redux.js.org/docs/basics/Reducers.html)
- Attempted to build a new Redux Action to add a todo item, but couldn't quite figure the best way to use the action to alter the existing state object
- Attempted to edit the props of the App component directly using a button I added to the jsx
- Decided to take a step back and get more first principles, using this tutorial: https://www.youtube.com/playlist?list=PL6gx4Cwl9DGBbSLZjvleMwldX8jGgXV6a

You can get a feel for the work I did here: https://github.com/downpat/todo-react/commit/a71af1d937f0dad8b8e95ef53998c9490587be1b

My initial reaction to React & Redux is that it's a proper software engineering solution to the state problems and performance issues that plague so many large javascript applications. When the initial tutorial started talking about the importance of not mutating state objects directly, I could see that React was taking a proper computer science approach to the the problem.

I'm fairly confident that I could pick this up and contribute, so let me know your thoughts and we can discuss further.

Thanks,
